src/ui/Omnibox/OmniboxAnalytics.ts

```diff
@@ -0,0 +1,70 @@
+import { compact, filter, map, reduceRight, last } from 'underscore';
+import { IAnalyticsOmniboxSuggestionMeta } from '../Analytics/AnalyticsActionListMeta';
+
+export class OmniboxAnalytics {
+  public partialQueries: string[];
+  public suggestionRanking: number;
+  public suggestions: string[];
+  public partialQuery: string;
+
+  constructor() {
+    this.partialQueries = [];
+    this.suggestions = [];
+  }
+
+  public buildCustomDataForPartialQueries(index?: number, suggestions?: string[]): IAnalyticsOmniboxSuggestionMeta {
+    this.partialQuery = last(this.partialQueries);
+    this.suggestions = suggestions || this.suggestions;
+    this.suggestionRanking = index || this.suggestionRanking;
+
+    return {
+      partialQueries: this.cleanCustomData(this.partialQueries),
+      suggestionRanking: this.suggestionRanking,
+      suggestions: this.cleanCustomData(this.suggestions),
+      partialQuery: this.partialQuery
+    };
+  }
+
+  public dumpInfo() {
+    console.log(this.partialQueries);
+  }
+
+  private cleanCustomData(toClean: string[], rejectLength = 256) {
+    // Filter out only consecutive values that are the identical
+    toClean = compact(
+      filter(toClean, (partial: string, pos?: number, array?: string[]) => {
+        return pos === 0 || partial !== array[pos - 1];
+      })
+    );
+
+    // Custom dimensions cannot be an array in analytics service: Send a string joined by ; instead.
+    // Need to replace ;
+    toClean = map(toClean, partial => {
+      return partial.replace(/;/g, '');
+    });
+
+    // Reduce right to get the last X words that adds to less then rejectLength
+    const reducedToRejectLengthOrLess = [];
+    reduceRight(
+      toClean,
+      (memo: number, partial: string) => {
```

Comment:
Could you extract that method to a separate private method?

Reply:
I can't pass `rejectLength` in the private function so no.

If this is really a problem, I could create a new class variable for it.

Personally, I find the function short and clear enough that I don't think it is necessary but if you feel different, just tell me and I'll do it.